src/main/java/org/apache/maven/plugins/checkstyle/Violation.java

```diff
@@ -0,0 +1,229 @@
+package org.apache.maven.plugins.checkstyle;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.Objects;
+import java.util.StringJoiner;
+
+/**
+ * Holds data about a single violation and represents the violation itself.
+ */
+public class Violation
+{
+
+  /**
+   * Indicates that a column is not set.
+   */
+  protected static final String NO_COLUMN = "-1";
+
+  /** The source file name relative to the project's root. */
+  private final String source;
+
+  /** File is the absolute file name to the checked file. */
+  private final String file;
+
+  private final String line;
+
+  private String column = NO_COLUMN;
```

Comment:
This is the only field that keeps these objects from being immutable. Could it be set in the constructor instead?

Reply:
I agree, let's make it immutable. the change is very little.
btw it is not so important,
I would like to move forward with this work, you have a new patch after this one, we can enhance it during the next step

Reply:
Please help me editing this commit in a way that I can use this as an additional (`@Nullable`)  constructor argument. I do not know where the checkstyle rules come from, or if you want to use `//CHECKSTYLE:OFF` (which does not work atm).

Reply:
Because setColumn is published API, we can't easily remove it after we publish it once. It would be better to not even include it here, and then add it if it's needed. But I tend to think we should fix this now.

Reply:
Yes, absolutely! Can you kindly edit my branch in a way that I can add another constructor argument?

Reply:
OK, l;et me take a stab at this.